allennlp/data/semparse/worlds/nlvr_world.py

```diff
@@ -0,0 +1,436 @@
+"""
+This module defines classes Object and Box (the two entities in the NLVR domain) and a NLVRWorld,
+which mainly contains an execution method and related helper methods.
+"""
+
+from collections import defaultdict
+import operator
+
+from typing import List, Dict, Set, Callable, Union, TypeVar
+
+import pyparsing
+
+
+AttributeType = Union[int, str]  # pylint: disable=invalid-name
+
+
+class Object:
+    """
+    ``Objects`` are the geometric shapes in the NLVR domain. They have values for attributes shape, color,
+    x_loc, y_loc and size. We take a dict read from the json file and store it here, and define a get method
+    for getting the attribute values. We need this to be hashable because need to make sets of ``Objects``
+    during execution, which get passed around between functions.
+
+    Parameters
+    ----------
+    object_dict : Dict[str, Union[str, int]]
+        The dict for each object from the json file.
+    """
+    def __init__(self, object_dict: Dict[str, AttributeType]) -> None:
+        self._object_dict: Dict[str, AttributeType] = {}
+        for key, value in object_dict.items():
+            if isinstance(value, str):
+                # The dataset has a hex code only for blue for some reason.
+                if key == "color" and value.startswith("#"):
+                    self._object_dict[key] = "blue"
+                else:
+                    self._object_dict[key] = value.lower()
+            else:
+                self._object_dict[key] = value
+
+    def get_attribute(self, attribute):
+        return self._object_dict[attribute]
+
+    def __str__(self):
+        color = self.get_attribute("color")
+        shape = self.get_attribute("type")
+        x_loc = self.get_attribute("x_loc")
+        y_loc = self.get_attribute("y_loc")
+        return "%s %s at (%d, %d)" % (color, shape, x_loc, y_loc)
+
+    def __hash__(self):
+        return hash(str(self))
+
+    def __eq__(self, other):
+        return str(self) == str(other)
+
+
+class Box:
+    """
+    This class represents each box containing objects in NLVR.
+
+    Parameters
+    ----------
+    name : str
+        This is just so that we can define a string representation to hash each object. It could be any unique
+        string.
+    objects_list : List[Dict]
+        List of objects in the box, as given by the json file.
+    """
+    def __init__(self, name: str, objects_list: List[Dict]) -> None:
+        self._name = name
+        self._objects_set = set([Object(object_dict) for object_dict in objects_list])
+
+    def get_objects(self) -> Set[Object]:
+        return self._objects_set
+
+    def __str__(self):
+        return self._name
+
+    def __hash__(self):
+        return hash(str(self))
+
+    def __eq__(self, other):
+        return str(self) == str(other)
+
+
+EntityType = TypeVar('EntityType', Object, Box)  # pylint: disable=invalid-name
+
+
+class NLVRWorld:
```

Comment:
`NlvrWorld`.  We use [Google's definition](https://google.github.io/styleguide/javaguide.html#s5.3-camel-case) of camel case.